### PR TITLE
SONAR-8857 make ws/qualityprofiles/search organization aware

### PR DESCRIPTION
**Pull request by bartfastiel**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileLookup.java [hunk 1]
```diff
@@ -71,27 +72,18 @@ private void incrementAncestors(QProfile profile, List<QProfile> ancestors, DbSe
...
 
-  private static List<QProfile> toQProfiles(List<QualityProfileDto> dtos) {
-    return newArrayList(Iterables.transform(dtos, ToQProfile.INSTANCE));
+  private static List<QProfile> toQProfiles(List<QualityProfileDto> dtos, OrganizationDto organization) {
+    return dtos.stream().map(dto -> QProfile.from(dto, organization)).collect(toList(dtos.size()));
   }
 
   @CheckForNull
   private QualityProfileDto findQualityProfile(String name, String language, DbSession session) {
     return db.qualityProfileDao().selectByNameAndLanguage(name, language, session);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileLookup.java]: MINOR Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QProfileFactory.java#L183](https://github.com/SonarSource/sonarqube/blob/02554b2fe04f9d4be20a0dae71a6b66e9c8b8d0c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L183): Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [QProfileFactory.java#L218](https://github.com/SonarSource/sonarqube/blob/02554b2fe04f9d4be20a0dae71a6b66e9c8b8d0c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L218): Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QualityProfileDao.java#L158](https://github.com/SonarSource/sonarqube/blob/02554b2fe04f9d4be20a0dae71a6b66e9c8b8d0c/server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L158): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QualityProfileDao.java#L171](https://github.com/SonarSource/sonarqube/blob/02554b2fe04f9d4be20a0dae71a6b66e9c8b8d0c/server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L171): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QProfileFactory.java#L168](https://github.com/SonarSource/sonarqube/blob/02554b2fe04f9d4be20a0dae71a6b66e9c8b8d0c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L168): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QProfileWsSupport.java#L79](https://github.com/SonarSource/sonarqube/blob/02554b2fe04f9d4be20a0dae71a6b66e9c8b8d0c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java#L79): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QProfileWsSupport.java#L98](https://github.com/SonarSource/sonarqube/blob/02554b2fe04f9d4be20a0dae71a6b66e9c8b8d0c/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java#L98): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml [hunk 2]
```diff
@@ -232,12 +233,11 @@
...
-    <where>
-      p.language in
-      <foreach collection="languages" open="(" close=")" item="language" separator=",">
-        #{language}
-      </foreach>
-    </where>
+    WHERE p.language in <foreach collection="languages" open="(" close=")" item="language" separator=",">#{language, jdbcType=VARCHAR}</foreach>
+    <if test="organizationUuid != null">
+      AND project.organization_uuid = #{organizationUuid, jdbcType=VARCHAR}
+      AND p.organization_uuid = #{organizationUuid, jdbcType=VARCHAR}
```
Comment (sns-seb) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml]: do not use an input twice, favor using a join which could give a hint the DB engine that the value is actually the same

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.sonar.api.utils.DateUtils.formatDateTime;
+import static org.sonar.db.component.ComponentTesting.newProjectDto;
+import static org.sonar.test.JsonAssert.assertJson;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_DEFAULTS;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_ORGANIZATION;
+import static org.sonarqube.ws.client.qualityprofile.QualityProfileWsParameters.PARAM_PROFILE_NAME;
+
+public class SearchActionTest2 {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java]: class is poorly name. I would expect a name that would tell me the difference between the two classes. It's definitely not obvious.

If it's only a matter of splitting the test methods over two classes, it's not a good reason. There should be a splitting logic (it could be technical, eg. one is mock based, the other is not). If I had to do maintenance on the `SearchAction` class I wouldn't know where to add unit tests.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  private DefaultOrganizationProvider defaultOrganizationProvider = TestDefaultOrganizationProvider.from(dbTester);
+  private QProfileWsSupport qProfileWsSupport = new QProfileWsSupport(dbTester.getDbClient(), userSession, defaultOrganizationProvider);
+  private ActiveRuleIndex activeRuleIndex = mock(ActiveRuleIndex.class);
+
+  private Language xoo1 = LanguageTesting.newLanguage("xoo1");
+  private Language xoo2 = LanguageTesting.newLanguage("xoo2");
+  private Languages languages = new Languages(xoo1, xoo2);
+
+  @Test
+  public void search_default_profile_by_profile_name_and_org() {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java]: the test setup is very long, I can't tell what make a difference from one QP to another. Some comment could help, some refactoring to highlight the important point too.

The consequence is: I can't tell whether the test is missing anything or doing anything wrong

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ */
+package org.sonar.server.qualityprofile.ws;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.sonar.server.exceptions.BadRequestException;
+import org.sonarqube.ws.client.qualityprofile.SearchWsRequest;
+
+import static org.junit.Assert.*;
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java]: bad wildcard import

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    SearchWsRequest request = new SearchWsRequest()
+            .setProfileName("bla")
+            .setProjectKey("blubb");
+
+    SearchDataLoader.validateRequest(request);
+  }
+
+  @Test
+  public void name_requires_either_component_or_defaults() throws Exception {
+    thrown.expect(BadRequestException.class);
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java]: if exception is thrown by `validateRequest`, `thrown` configuration should be done right before `validateRequest` is called

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v64/DbVersion64.java [hunk 7]
```diff
@@ -32,7 +32,11 @@ public void addSteps(MigrationStepRegistry registry) {
       .add(1602, "Add RULES_PROFILES.ORGANIZATION_UUID", AddQualityProfileOrganizationUuid.class)
       .add(1603, "Set RULES_PROFILES.ORGANIZATION_UUID to default", SetQualityProfileOrganizationUuidToDefault.class)
       .add(1604, "Make RULES_PROFILES.ORGANIZATION_UUID not nullable", MakeQualityProfileOrganizationUuidNotNullable.class)
+
       .add(1605, "Drop unique index on RULES_PROFILES.KEE", DropUniqueIndexOnQualityProfileKey.class)
-      .add(1606, "Make RULES_PROFILES.ORGANIZATION_UUID and KEE unique", MakeQualityProfileOrganizationUuidAndKeyUnique.class);
+      .add(1606, "Make RULES_PROFILES.ORGANIZATION_UUID and KEE unique", MakeQualityProfileOrganizationUuidAndKeyUnique.class)
+      .add(1607, "Drop unique index on RULES_PROFILES.ORGANIZATION_UUID and KEE", DropUniqueIndexOnQualityProfileOrganizationUuidAndKey.class)
+      .add(1608, "Make RULES_PROFILES.KEE unique", MakeQualityProfileKeyUnique.class)
```
Comment (sns-seb) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v64/DbVersion64.java]: don't forget to add yourself a task to merge these migrations with 1605 and 1606

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java [hunk 8]
```diff
@@ -183,7 +184,7 @@ public void search_for_language() throws Exception {
   @Test
   public void search_for_project_qp() {
     long time = DateUtils.parseDateTime("2016-12-22T19:10:03+0100").getTime();
-    OrganizationDto org = db.organizations().insert();
+    OrganizationDto org = db.getDbClient().organizationDao().selectByKey(db.getSession(), defaultOrganizationProvider.get().getKey()).get();
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java]: you can use `db.getDefaultOrganisation()` instead of all this

File: server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml [hunk 9]
```diff
@@ -237,6 +239,8 @@
       <foreach collection="languages" open="(" close=")" item="language" separator=",">
         #{language}
       </foreach>
+      AND p.organization_uuid = #{organizationUuid}
```
Comment (sns-seb) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml]: jdbcType is missing from this commit. They are incorrecly added in the unrelated next commit

File: server/sonar-db-dao/src/test/resources/org/sonar/db/qualityprofile/QualityProfileDaoTest/delete-result.xml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 <dataset>
 
-  <rules_profiles id="2" name="Sonar Way" language="js" organization_uuid="org-123" parent_kee="[null]" kee="js_sonar_way" is_default="[false]"
```
Comment (sns-seb) [inline server/sonar-db-dao/src/test/resources/org/sonar/db/qualityprofile/QualityProfileDaoTest/delete-result.xml]: I don't understand how this change is necessary
any name for organization is good
imho, the new one is uselessly too long

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfile.java [hunk 11]
```diff
@@ -23,11 +23,13 @@
 import javax.annotation.Nullable;
 import org.apache.commons.lang.builder.ReflectionToStringBuilder;
 import org.apache.commons.lang.builder.ToStringStyle;
+import org.sonar.db.organization.OrganizationDto;
 import org.sonar.db.qualityprofile.QualityProfileDto;
 
 public class QProfile {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfile.java]: I guess this class was introduced to cut the link and dependency with the DTO object
Seems very odd to add a DTO back as a field

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java [hunk 12]
```diff
@@ -221,30 +234,15 @@ private static boolean hasLanguage(SearchWsRequest request) {
...
-      return new QProfile()
-        .setKey(input.getKey())
-        .setName(input.getName())
-        .setLanguage(input.getLanguage())
-        .setDefault(input.isDefault())
-        .setRulesUpdatedAt(input.getRulesUpdatedAt())
-        .setLastUsed(input.getLastUsed())
-        .setUserUpdatedAt(input.getUserUpdatedAt());
-    }
+  public static QProfile qualityProfileDtoToQProfile(QualityProfileDto input, OrganizationDto organization) {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java]: how does this method compare to `QProfile.from`?
seems to me this one just doesn't copy the parent and id fields, which does look like a bug

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfilesWsTest.java [hunk 13]
```diff
@@ -129,7 +129,7 @@ public void define_search() {
     WebService.Action search = controller.action("search");
     assertThat(search).isNotNull();
     assertThat(search.isPost()).isFalse();
-    assertThat(search.params()).hasSize(4);
+    assertThat(search.params()).hasSize(5);
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfilesWsTest.java]: no assertion on new param itself?

File: server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml [hunk 14]
```diff
@@ -55,6 +55,9 @@
     SELECT
     <include refid="profilesColumns"/>
     FROM rules_profiles p
+    <if test="organizationUuid != null">
```
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml]: Done. :honeybee:

File: server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml [hunk 15]
```diff
@@ -141,13 +144,11 @@
...
-    <where>
-      p.is_default=${_true}
-      AND p.language in
-      <foreach collection="languages" open="(" close=")" item="language" separator=",">
-        #{language}
-      </foreach>
-    </where>
+    WHERE p.is_default=${_true}
+    AND p.language in <foreach collection="languages" open="(" close=")" item="language" separator=",">#{language, jdbcType=VARCHAR}</foreach>
+    <if test="organizationUuid != null">
```
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml]: Done. :ant:

File: server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml [hunk 16]
```diff
@@ -232,12 +233,11 @@
...
     JOIN project_qprofiles pp ON pp.profile_key=p.kee
     JOIN projects project ON pp.project_uuid=project.uuid AND project.kee=#{projectKey}
-    <where>
-      p.language in
-      <foreach collection="languages" open="(" close=")" item="language" separator=",">
-        #{language}
-      </foreach>
-    </where>
+    WHERE p.language in <foreach collection="languages" open="(" close=")" item="language" separator=",">#{language, jdbcType=VARCHAR}</foreach>
+    <if test="organizationUuid != null">
```
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml]: Done. :turtle:

File: server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml (hunk 2 above)
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/qualityprofile/QualityProfileMapper.xml]: Done. :+1:

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java (hunk 3 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java]: Valid point. I merged the two tests.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java (hunk 4 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest2.java]: Not easy to do, but the test is now better readable.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java (hunk 5 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java]: Done.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java (hunk 6 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchDataLoaderTest.java]: Done.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java (hunk 8 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java]: Done.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java (hunk 12 above)
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java]: Fixed. :star2:

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfilesWsTest.java (hunk 13 above)
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/QProfilesWsTest.java]: I added a test and found a :bug:!

Fixed.⚡️

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 5 info

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QProfileFactory.java#L183](https://github.com/SonarSource/sonarqube/blob/f460d0e328ad63f841dfd947f2d4a3923536f0aa/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L183): Remove this use of "selectByNameAndLanguage"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QualityProfileDao.java#L158](https://github.com/SonarSource/sonarqube/blob/f460d0e328ad63f841dfd947f2d4a3923536f0aa/server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L158): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QualityProfileDao.java#L171](https://github.com/SonarSource/sonarqube/blob/f460d0e328ad63f841dfd947f2d4a3923536f0aa/server/sonar-db-dao/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java#L171): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QProfileFactory.java#L168](https://github.com/SonarSource/sonarqube/blob/f460d0e328ad63f841dfd947f2d4a3923536f0aa/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileFactory.java#L168): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QProfileWsSupport.java#L79](https://github.com/SonarSource/sonarqube/blob/f460d0e328ad63f841dfd947f2d4a3923536f0aa/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java#L79): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QProfileWsSupport.java#L98](https://github.com/SonarSource/sonarqube/blob/f460d0e328ad63f841dfd947f2d4a3923536f0aa/server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileWsSupport.java#L98): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'
